### ISSUE TITLE
Retain CSS Styling When Pasting Into Gutenberg

### ISSUE THREAD

**Issue by davidshq** [WordPress/gutenberg#18381]
When pasting into a Gutenberg blog (e.g., paragraph) Gutenberg retains only semantic markup and strips out CSS. Seems from https://github.com/WordPress/gutenberg/issues/15590#issuecomment-523794936 this is the expected behavior.

I think I understand the reasoning behind retaining only semantic markup - if CSS was included when pasting from sites it could result in some pretty confusing/odd results. However, this leaves users who create their documents in another application (Word, Google Docs, etc.) in a lurch as all their hard work formatting the style of the document is lost.

And I imagine there is some concern about cluttering the UI with too many options, but it seems to me that an option to retain CSS could be retained without cluttering the UI. Either:

1. When pasting a user could use a keyboard shortcut to retain the CSS.
2. When first pasting the user would see a modal asking how they want to paste (as semantic, sematic+css, or plain text) and allowing them to make a default selection.

Would like to hear more about why CSS is being stripped, any workarounds, and what might be done in the future.

**Comment by ellatrix** [comment]
Even if we did preserve some styling, how would it be edited?
Could you give some examples of styles you wish to keep?

I'm open to preserving some stuff on a case by case basis, but not to retaining all CSS styling.
E.g. we could keep some font size on certain blocks because we support it in Gutenberg.

**Comment by davidshq** [comment]
Hi Ella,

I'm not sure what the best way would be...my thought would be inline styling. Which I'm not sure matches up with the "Gutenberg way" but pragmatically would be very useful (imho).

Right now when I paste a Google doc's contents into the Gutenberg editor I end up viewing the HTML for each block and making manual tweaks to the code to re-add the styling that is lost.

[Here is an example of a Google Doc](https://docs.google.com/document/d/1yu8nAoiQdiLg7RIckSWkb1M2OjFJQH6qA69nyNeuWks/edit?usp=sharing) that I recently pasted into the Gutenberg editor.

And here is the [page that (after manual editing of the HTML) resulted](https://liquidchurch.com/blog/notes/spend-less/).

For my particular use case there are only a few stylings that need to be retained:
1. Text Alignment
2. Font Size
3. Font Color

To see these supported would be huge...

In the short-term, I'd love a way to override the default pasteHandler (particularly on richText). Maybe there is a way, but I haven't been able to figure it out, [nor has anyone else I've asked](https://wordpress.stackexchange.com/questions/352134/how-would-one-modify-the-filtering-gutenberg-applies-to-pasted-content). Do you have any suggestions?

Right now I can edit the code manually for the pages (created on a weekly basis) but the plugin is meant to be utilized by non-technical individuals and editing code isn't ideal...The [plugin is open source](https://github.com/LiquidChurch/lqd-notes), so I'm hoping it could be of use to teachers, speakers, etc. beyond the church context -- and generally I would expect these to be non-technical users as well.

**Comment by ellatrix** [comment]
There are many cases where the styling isn't wanted. It's not easy to distinguish between these use cases. :/

**Comment by davidshq** [comment]
Yes, I imagine most individuals copying from a web page and into WP would not want to retain the original site's styling. 

I'd suggest a keyboard shortcut would be sufficient to allow folks to paste in formatted content. A few potentials:
- Ctrl+Shift+F (F for retain formatting, similar to Ctrl+Shift+V for removing formatting)
- Alt+Shift+V (allows retaining the familiar V)

Alternatively, if there is an easy way to override the pasteHandler it would be pretty easy for non-WP core developers to add a setting that allowed users to select whether they wanted the default paste for specific blocks to be formatted or unformatted (and even which tags/styles to retain).

**Comment by sundoulos2** [comment]
> Even if we did preserve some styling, how would it be edited?
> Could you give some examples of styles you wish to keep?
> 
> I'm open to preserving some stuff on a case by case basis, but not to retaining all CSS styling.
> E.g. we could keep some font size on certain blocks because we support it in Gutenberg.

Here's my workaround:

I create my "post" in a rich text editor (it's primary use is for something else first, then posting to my blog). I copy the document to an online html editor, in my case Summernote. I switch to code view, copy the code and paste it into Gutenberg's code editor. When I exit the code editor, my post has the same appearance as the original document.

iirc, there used to be two toolbar options in the web-based editor: paste as text only or paste in Word format. Maybe bring that feature (or one similar to it) back?

**Comment by stronenv** [comment]
Plus one for this. Especially for inline CSS we have the equivalent and block controls for (and inline CSS), like text-alignment, color, background-color, font-size, etc. 

For example: 
```
<p style="color:#f7f273;background-color:#4a4a4a;font-size:24px;text-align:center;">The quick brown fox jumps over the lazy dog</p>
```
Becomes:
```
<!-- wp:paragraph {"align":"center","style":{"elements":{"link":{"color":{"text":"#f7f273"}}},"typography":{"fontSize":"24px"},"color":{"text":"#f7f273","background":"#4a4a4a"}}} -->
<p class="has-text-align-center has-text-color has-background has-link-color" style="color:#f7f273;background-color:#4a4a4a;font-size:24px">The quick brown fox jumps over the lazy dog</p>
<!-- /wp:paragraph -->
```
